crates/ruff_diagnostics/src/fix.rs

```diff
@@ -0,0 +1,43 @@
+use rustpython_parser::ast::Location;
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+use crate::edit::Edit;
+
+/// A collection of edits to be applied to a source file.
+#[derive(Default, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct Fix {
+    edits: Vec<Edit>,
+}
+
+impl Fix {
+    pub fn new(edits: Vec<Edit>) -> Self {
+        // TODO(charlie): Ensure that the edits are sorted and non-overlapping.
+        debug_assert!(!edits.is_empty(), "Fix must have at least one edit");
```

Comment:
There's some awkwardness to this struct... I wish these constraints were expressed in the type system (or, at least, a subset of them). It'd also be nice if we didn't require a `Vec` here at all. The vast majority of fixes will have exactly one edit. Others, in practice, will have a bounded number of edits (maybe two? three?).

We could use `smallvec` to avoid heap allocations. I considered instead using an enum, like:

```rust
pub enum Fix {
  One(Edit),
  Two(Edit, Edit),
  ...
```

...and so on.

Would love feedback, this feels like a good learning opportunity for me.

Reply:
What's the motivation for enforcing that the edits are not empty? Does it simplify the consumption of edits downstream? 

I'm asking because `Fix::none` and the derived `Default` implementation generate empty fixes and the API provides an `is_none` function, making me believe this invariant isn't necessary (other than accidental creation of `Fix`es without any edits).

We would need to profile to know whether it makes sense to use smallvec or an enum with `Empty`, `One` and `Many` variants. My intuition is that it shouldn't matter much for projects that have adopted Ruff because they normally have most issues fixed (a few issues from the now-edited files and ignored suppressions). It may be different for projects that run ruff the first time but using an enum instead of a `Vec` has additional costs when reading the values (an iterator needs to determine the right variant in every `next` call) and may be less optimized. This means we would be trading one cost for another. 

I think it could be more interesting to use a single edits `Vec` for all changes in a file and uses indices into that vec to get the edits of a single fix. But again, I don't this is necessary just now.

Reply:
All makes sense.

I think I'm getting hung up on the `Option<Vec<...>>` thing, whereby using an empty `Vec` is seen as preferable to using `None`. Because of that convention, at the level above, I was going to use ` Fix { edits: vec![] }` to represent "no fix", rather than `None`. But that creates some awkwardness in the API.